### ISSUE TITLE
Bad Request

### ISSUE THREAD

**Issue by minohimself** [SevenSpikes/nopCommerce-Api-SampleApplication#2]
Hi guys, im using nop3.8 and followed the steps and getting Bad Request error.

What could be wrong ?

**Comment by poyker** [comment]
@minohimself 

It could be everything. 
On which step did you get this error?

**Comment by RobertCarlaw** [comment]
Hi,

I've managed to install the plugin into Nop (3.90), I've created a Api client in the nop admin at: /Plugins/ManageClientsAdmin/

Now using the Sample application I've entered the details, copying them from the nop admin.

Whenever I hit 'GetAccessToken' from the sample application I get a 400 bad request. Which is failing on
OAuthController.Authorize  on the If statement checking for a 200.

I've attached the following images which may help. I've also tried removing slashes and running through IIS but still get the same error.

Any ideas?

![api client nop](https://cloud.githubusercontent.com/assets/16261461/25138761/49d63388-2453-11e7-9571-2630c585b9ea.jpg)
![sample app](https://cloud.githubusercontent.com/assets/16261461/25138760/49d23256-2453-11e7-8509-bb4a6bc45677.jpg)

**Comment by RobertCarlaw** [comment]
Ok it was a config error, The callback URL in nop needs to have '/token' at the end.
Once I'd made this change it all works.

**Comment by DarshitThakkar23** [comment]
When i try to send webRequest using AuthorizeClient method i get 400 bad request error. i have attached following images. Please help. Thanks.
![image1](https://user-images.githubusercontent.com/29428551/27129950-d3df5b40-5121-11e7-8db5-a9a559cab2c7.png)
![image2](https://user-images.githubusercontent.com/29428551/27129958-d97ef8da-5121-11e7-84fa-e29b49b48080.png)
![image3](https://user-images.githubusercontent.com/29428551/27129962-dc4c4c16-5121-11e7-8e9d-f77dd40c13df.png)

**Comment by RobertCarlaw** [comment]
Not sure, your config looks correct. 

- make sure the Api Client is active 

Put a break point in the NopCommerce plugin - Nop.Plugin.Api.Controllers.OAuth.OAuthController.Authorize method see if it's hitting it.

**Comment by DarshitThakkar23** [comment]
Thanks for the reply. Now i am able to access get customers method. just restart the vs2015, clean and rebuilt solution(NopCommerce) and it works.

**Comment by DarshitThakkar23** [comment]
I am not able to get RoleIds when i use GetCustomers web api method. It always show as Count = 0. Is there anythig i need to configure?

**Comment by RobertCarlaw** [comment]
Hi,

I'd put a break point and step through, it might not be in the DTO
mappings, so you may need to add it. Other things to check that the
customer is active and it's not data related.

**Comment by lculjak** [comment]
is it possible that after sometime inactive I get /oauth/authorize response  BadRequest 'Bad Request' ?

if I deploy the website, the error is fixed, but then is back.

**Comment by RobertCarlaw** [comment]
HI,

Not 100% sure, After a period of time the access token will expire, I don't think it should return a bad request response.  
Have you implemented the refresh token?

**Comment by poyker** [comment]
@RobertCarlaw 
Yes, the refresh token is implemented and there is sample code in the [Sample Application](https://github.com/SevenSpikes/nopCommerce-Api-SampleApplication) that shows you how to obtain a refresh token.

**Comment by lculjak** [comment]
I am using the snipet sugested in https://github.com/SevenSpikes/api-plugin-for-nopcommerce/issues/31
is there any issue to get a new token every time that I need to call the API ?
I understand that it may have some performance degradation, but in this case it is not an issue.

**Comment by poyker** [comment]
@lculjak 

Don't see any issues but if you already have a token then there is no real need to obtain a new one until the current one is valid. So you should get a new token (preferably using the refresh token) only when the current access token is expired in which case you will get 401 Unauthorized response when trying to access any of the endpoints.

**Comment by lculjak** [comment]
I think that I found the reason for the Bad Request status code. My callback url is http://localhost:19890/token but using RestSharp I don't need to setup that url.
If I setup call back url to http://localhost:19890/ it is working. I will continue checking if there is any error later.

**Comment by archie7484** [comment]
I don't know what's going on, I'm in the same problem and this is getting hopeless. 

This is the setup:

![response0](https://user-images.githubusercontent.com/35822864/35479945-8daa7080-043e-11e8-8376-7d3f4f70cde8.png)

This is the response inside nopCommerce:

![response1](https://user-images.githubusercontent.com/35822864/35479949-96db1678-043e-11e8-98da-ecd5353d56b2.png)

Anybody Please help...

**Comment by poyker** [comment]
@archie7484 

You should remove the /en at the end of the Server URL.
